File: problems/morra/services/python/compute_hash.py
#!/usr/bin/env python3

from os import environ
from sys import exit

ENV_lang = environ.get("TAL_lang")
ENV_s = environ.get("TAL_white_string")

def rabin_karp(ASCII_string):
    hash = 0
    for i in range(len(ASCII_string)):
        hash = ( hash * 257 + ord(ASCII_string[i]) ) & ((1 << 64) -1)
    return hash

if ENV_s != None:
    print(f"h({ENV_s})=\n{rabin_karp(ENV_s)}")
else:
    print("Poichè non hai specificato il parametro 'clean_string', ti chiediamo di immettere ora la stringa in chiaro, di cui computare l'hash:");
    str=input()
    print(f"h({str})=\n{rabin_karp(str)}")
    
exit(0)

File: problems/morra/services/c/compute_hash.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <inttypes.h>

uint64_t rabin_karp(const char* ASCII_string) {
    uint64_t hash = 0;
    size_t size = strlen(ASCII_string);
    for(size_t i = 0; i < size; i++) {
        hash = hash * 257ULL + ASCII_string[i];
    }
    return hash;
}

int main() {
   char* ENV_s = getenv("TAL_clean_string");
   if(ENV_s!=NULL) {
     printf("h(%s)=\n", ENV_s);
     uint64_t h = rabin_karp(ENV_s);
     printf("%" PRIu64 "\n", h);
   }
   else {
     char str[100];
     printf("Poichè non hai specificato il parametro 'clean_string', ti chiediamo di immettere ora la stringa in chiaro, di cui computare l'hash:\n");
     // scanf("%s",str);  but safer w.r.t. possible attacks:
     fgets(str,100,stdin);
     if(str[strlen(str)-1] == '\n')
       str[strlen(str)-1] = '\0';
     if(strlen(str)>64) {
       fprintf(stderr, "Hai immesso una stringa più lunga di 64 caratteri.\n");
       exit(0);
     }
     printf("h(%s)=\n", str);
     uint64_t h = rabin_karp(str);
     printf("%" PRIu64 "\n", h);
   }
   exit(0);
}
